vowpalwabbit/core/src/reductions/cb/cb_explore_adf_large_action_space.cc

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "vw/core/reductions/cb/cb_explore_adf_large_action_space.h"
+
+#include "details/large_action_space.h"
+#include "vw/config/options.h"
+#include "vw/core/gd_predict.h"
+#include "vw/core/label_parser.h"
+#include "vw/core/rand_state.h"
+#include "vw/core/reductions/cb/cb_adf.h"
+#include "vw/core/reductions/cb/cb_explore.h"
+#include "vw/core/setup_base.h"
+#include "vw/explore/explore.h"
+
+#include <algorithm>
+#include <cmath>
+#include <functional>
+#include <vector>
+using namespace VW::cb_explore_adf;
+
+namespace VW
+{
+namespace cb_explore_adf
+{
+void cb_explore_adf_large_action_space::predict(VW::LEARNER::multi_learner& base, multi_ex& examples)
+{
+  predict_or_learn_impl<false>(base, examples);
+}
+void cb_explore_adf_large_action_space::learn(VW::LEARNER::multi_learner& base, multi_ex& examples)
+{
+  predict_or_learn_impl<true>(base, examples);
+}
+
+cb_explore_adf_large_action_space::cb_explore_adf_large_action_space(uint64_t d_, float gamma_, VW::workspace* all_)
+    : d(d_), gamma(gamma_), all(all_), seed(all->get_random_state()->get_current_state())
+{
+}
+
+void cb_explore_adf_large_action_space::calculate_shrink_factor(const ACTION_SCORE::action_scores& preds, float min_ck)
+{
+  shrink_factors.clear();
+  for (size_t i = 0; i < preds.size(); i++)
+  { shrink_factors.push_back(std::sqrt(1 + d + (gamma / 4.0f * d) * (preds[i].score - min_ck))); }
+}
+
+inline void just_add_weights(float& p, float, float fw) { p += fw; }
+
+void cb_explore_adf_large_action_space::generate_Q(const multi_ex& examples)
```

Comment:
Does this produce the same result if called twice? If `Q` is not in the default state?

Reply:
Yes it does, what is the default state of Q? This function resizes Q and re populates its cells

Reply:
Whatever state the default constructor places it in when the object is created

Reply:
Yeah the function shouldn't care what state Q is in, it will take care of it

Reply:
can add it as a test though

Reply:
Only if you want to, resize and fill makes sense to me